### PR TITLE
don't encode only locally used attrs

### PR DESCRIPTION
**Pull request by lcnr**
Part of https://github.com/rust-lang/compiler-team/issues/505.

We now filter builtin attributes before encoding them in the crate metadata in case they should only be used in the local crate. To prevent accidental misuse `get_attrs` now requires the caller to state which attribute they are interested in. For places where that isn't trivially possible, I've added a method `fn get_attrs_unchecked` which I intend to remove in a followup PR.

After this pull request landed, we can then slowly move all attributes to only be used in the local crate while being certain that we don't accidentally try to access them from extern crates.

cc https://github.com/rust-lang/rust/pull/94963#issuecomment-1082924289

### CODE REVIEW

**Comment by rylev** [comment]
Unfortunately, I believe the performance gains seen here are not real. The benchmark showing such large improvements was switched out for another version that uses a different feature set and compiles in general roughly twice as fast. You can find the discussion about this [here](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/Time.20for.20secondary.20benchmarks/near/277389481)

**Comment by davidtwco** [comment]
r? @davidtwco

Let's trigger another perf run to see what the actual improvements are (excluding the big improvements from the last run, probably going to be a modest improvement).

@bors try @rust-timer queue

**Review by davidtwco — APPROVED** [review]
LGTM, r=me if perf results are good and you want to land this

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (bjorn3) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1351]: Does this keep non-builtin attributes like registered with `#![register_tool]`?

File: compiler/rustc_middle/src/ty/mod.rs
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/mod.rs:2175]: There is a grammar issue here I think.

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (lcnr) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1351]: yes, it only removes `builtin` attributes where `only_local` is true.

Comment (bjorn3) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1351]: :+1:

**Comment by bjorn3** [comment]
Could we have a debug assertion to check that there is no attempt made to get an `only_local` attribute from a foreign crate?

File: compiler/rustc_feature/src/builtin_attrs.rs
Comment (bjorn3) [inline compiler/rustc_feature/src/builtin_attrs.rs:382]: This can be skipped too I think.

**Comment by lcnr** [comment]
@bors rollup=never

This now pretty much always checks that we don't access `only_local` attributes except in a few places in rustdoc and when dealing with `weak_lang_items`. These aren't completely trivial so I didn't update these for now.

Intend to make further progress on https://github.com/rust-lang/compiler-team/issues/505 in followup PRs but want to get this merged as it is quite prone to merge conflicts.

@davidtwco the new commits change `get_attrs` to require the user to eagerly specify which attribute they're interested in. This allows us to error if the required attribute is `only_local`. Apart from this nothing changed here.

One more perf run to make sure I didn't break anything.

@bors try @rust-timer queue

File: compiler/rustc_attr/src/builtin.rs
Comment (lcnr) [inline compiler/rustc_attr/src/builtin.rs:863]:
```suggestion
    assert!(attr.has_name(sym::repr), "expected `#[repr(..)]`, found: {:?}", attr);
```
before merging. Don't want to do it during the perf run.

**Comment by lcnr** [comment]
I most of the slowdown comes from my changes to `codegen_fn_attrs`. Going to cache that query in the crate metadata which allows us to mostly keep the current impl.

**Review by davidtwco — APPROVED** [review]

**Comment by lcnr** [comment]
@bors try @rust-timer queue

let's the perf impact with `inline` changed to `only_local` and instead adding `codegen_fn_attrs` to the crate metadata

**Comment by lcnr** [comment]
rebased this on top of #96473 for perf testin

@bors try @rust-timer queue

**Review by davidtwco — APPROVED** [review]

**Comment by davidtwco** [comment]
r=me if you're done improving this :)

File: compiler/rustc_middle/src/ty/mod.rs
Comment (BGR360) [inline compiler/rustc_middle/src/ty/mod.rs:2194]: Hey @lcnr , if and when you get around to removing this, what will be the recommended way to query for multi-component attributes?

We have a compiler tool with a late lint pass that checks for the presence of some `foo::bar` attribute on items, and it seems like if you get rid of this function, I won't be able to query for that since `PartialEq<Symbol> for Path` wants the path to have exactly one component.

Comment (BGR360) [reply compiler/rustc_middle/src/ty/mod.rs:2194]: I don't think so. If I `get_attrs` with `attr = Symbol("foo")`, then that calls `a.has_name("foo")` for each attribute, which does `a.path == "foo"`, and that `==` wants `path` to have exactly one component:

https://github.com/rust-lang/rust/blob/ebf95836e3cfc8c442cc71239324d947aafa3eac/compiler/rustc_ast/src/ast.rs#L103-L108

Comment (lcnr) [reply compiler/rustc_middle/src/ty/mod.rs:2194]: ah... so we should maybe add something like `get_tool_attrs` or something? :thinking: we definitely want to keep supporting this in some way '^^

Comment (BGR360) [reply compiler/rustc_middle/src/ty/mod.rs:2194]: Something like that, yeah.

**Pull request merged**
